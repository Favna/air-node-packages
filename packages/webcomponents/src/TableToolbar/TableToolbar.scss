@import '../styles/vars';

.heightCorrection {
  flex-basis: 0%;
  height: 4rem;
}

.toolbar {
  @extend %beautifulCenter;
  background-color: $lightGrey;
  display: flex;
  height: 4rem;
  margin-top: 5vh;
  width: 100%;
}

.searchFieldContent {
  border: 1px solid $mediumGrey;
  border-radius: 0;
  box-shadow: none;
  color: $grey;
  font-size: $fontSizeReset;
  margin: 0 auto;
  margin-left: 24vw;
  width: 20vw;
}

.searchIconButton {
  color: $blue;
  opacity: 1;
}

// stylelint-disable-next-line selector-max-compound-selectors
.ie11SearchBarTextCorrection > div > div > input {
  min-height: 2.5rem;
}
