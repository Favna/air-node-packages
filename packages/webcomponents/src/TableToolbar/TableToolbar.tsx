import Grid from '@material-ui/core/Grid/Grid';
import classnames from 'classnames';
import debouncer from 'lodash.debounce';
import React, { FC, Fragment } from 'react';
import { customCss, dataQa, label } from '../constants';
import SearchBar, { SearchBarProps } from '../SearchBar/SearchBar';
import css from './TableToolbar.scss';
import PaperButton from '../PaperButton/PaperButton';
import { PaperButtonProps } from '../PaperButton';

export interface RenderIconProps {
  /** The Material Icon to display in this paper button */
  icon: JSX.Element;
  /** The function to trigger when clicking this button */
  clickEvent: PaperButtonProps['onClick'];
  /**
   * The text content to put in the tooltip
   * @default 'Click Me'
   */
  tooltipText?: PaperButtonProps['tooltipText'];
  /**
   * The placement for the tooltip
   * @default top
   */
  tooltipPlacement?: PaperButtonProps['tooltipPlacement'];
  /**
   * Disables the inclusion of a tooltip
   * @default false
   */
  disableTooltip?: PaperButtonProps['disableTooltip'];
}

export interface TableToolbarProps extends Pick<SearchBarProps, 'paperElevation'> {
  /** The debounce timeout to wait until a search action should be performed */
  searchdebounce?: number;
  /** The label to display when no text is in the search input box */
  searchplaceholderlabel: label;
  /** The action to trigger when searching (taking debounce into account) */
  onsearchinput: (searchTerms: string) => unknown;
  /** The action to trigger when clearing the search */
  onsearchclear: () => unknown;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
  /** Custom CSS classes to pass to the toolbar */
  customclasses?: customCss;
  /** Custom CSS classes to pass to the inner searchbar */
  customSearchbarClasses?: customCss;
  /** Any additional icons to display before the search bar */
  extraIcons?: RenderIconProps[];
}

export const renderIcons = (iconData: RenderIconProps[]) => {
  return (
    iconData.map((data, index) => (
      <Grid item key={`icon-${index}`}>
        <PaperButton
          icon={data.icon}
          onClick={data.clickEvent}
          tooltipText={data.tooltipText}
          tooltipPlacement={data.tooltipPlacement || 'top'}
          disableTooltip={data.disableTooltip}
        />
      </Grid>
    ))
  );
};

const TableToolbar: FC<TableToolbarProps> = props => {
  const debouncedSearch = debouncer((input: string) => props.onsearchinput(input), props.searchdebounce || 400);

  return (
    <div className={classnames(css.toolbar, props.customclasses)} data-qa={props['data-qa']}>
      <Grid container direction='row' justify='flex-end' alignItems='center' spacing={2}>
        {
          props.extraIcons && props.extraIcons.length
            ? renderIcons(props.extraIcons)
            : <Fragment />
        }
        <Grid item key={2}>
          <SearchBar
            data-qa='table-search-bar'
            placeholder={`${props.searchplaceholderlabel}...`}
            onChange={e => debouncedSearch(e.target.value)}
            onCancelSearch={props.onsearchclear}
            className={classnames(css.searchFieldContent, css.ie11SearchBarTextCorrection, props.customSearchbarClasses)}
            paperElevation={props.paperElevation}
          />
        </Grid>
        <Grid item key={1} />
      </Grid>
    </div>
  );
};

export default TableToolbar;