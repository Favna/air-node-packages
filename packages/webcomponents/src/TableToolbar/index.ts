export { default as TableToolbar, TableToolbarProps, RenderIconProps } from './TableToolbar';