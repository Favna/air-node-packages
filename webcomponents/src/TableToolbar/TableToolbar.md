### Usage of the TableToolbar component:

**Examples:**

```tsx
import React from 'react';
import { TableToolbar } from '@rws-air/webcomponents';

<TableToolbar
      searchPlaceholderLabel={'Search...'}
      onSearchInput={() => console.log('Got some search input')}
      onSearchClear={() => console.log('Cleared the search')}
/>
```

```jsx
import React from 'react';
import { TableToolbar } from '@rws-air/webcomponents';

<TableToolbar
      searchPlaceholderLabel={'Search...'}
      onSearchInput={() => console.log('Got some search input')}
      onSearchClear={() => console.log('Cleared the search')} 
/>
```

