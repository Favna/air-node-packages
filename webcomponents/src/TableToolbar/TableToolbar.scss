@import '../styles/vars.scss';

.heightCorrection {
  flex-basis: 0%;
  height: 4rem;
}

.toolbar {
  @include beautifulCenter;
  background-color: $light-grey;
  display: flex;
  height: 4rem;
  margin-top: 5vh;
  width: 100%;
}

.searchFieldContent {
  border: 1px solid $medium-grey;
  border-radius: 0;

  box-shadow: none;
  color: $grey;
  font-size: $font-size-reset;
  margin: 0 auto;
  margin-left: 24vw;
  width: 20vw;
}

.searchIconButton {
  color: $blue;
  opacity: 1;
}

.ie11-searchBarTextCorrection > div > div > input {
  min-height: 2.5rem;
}
