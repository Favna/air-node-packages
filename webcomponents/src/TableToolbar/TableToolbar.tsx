import Grid from '@material-ui/core/Grid/Grid';
import SearchIcon from '@material-ui/icons/Search';
import classnames from 'classnames';
import debouncer from 'lodash.debounce';
import React, { FC } from 'react';
import SearchBar, { SearchBarProps } from '../SearchBar/SearchBar';
import css from './TableToolbar.scss';
import { customCss, dataQa, label } from 'typings';

export type TableToolbarProps = {
  /** The debounce timeout to wait until a search action should be performed */
  searchdebounce?: number;
  /** The label to display when no text is in the search input box */
  searchplaceholderlabel: label;
  /** The action to trigger when searching (taking debounce into account) */
  onsearchinput: (searchTerms: string) => unknown;
  /** The action to trigger when clearing the search */
  onsearchclear: () => unknown;
  /** data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
  /** Custom CSS classes to pass to the toolbar */
  customclasses?: customCss;
  /** Custom CSS classes to pass to the inner searchbar */
  customSearchbarClasses?: customCss;
} & Pick<SearchBarProps, 'paperElevation'>;

const TableToolbar: FC<TableToolbarProps> = props => {
  const debouncedSearch = debouncer((input: string) => props.onsearchinput(input), props.searchdebounce || 400);

  return (
    <div className={classnames(css.toolbar, props.customclasses)} data-qa={props['data-qa']}>
      <Grid container direction='row' justify='space-between' alignItems='center'>
        <Grid item key={1} xs={6} className={css.heightCorrection} />
        <Grid item key={2} xs={6}>
          <SearchBar
            data-qa='table-search-bar'
            placeholder={`${props.searchplaceholderlabel}...`}
            onChange={e => debouncedSearch(e.target.value)}
            onCancelSearch={props.onsearchclear}
            className={classnames(css.searchFieldContent, css['ie11-searchBarTextCorrection'], props.customSearchbarClasses)}
            searchIcon={<SearchIcon className={classnames(css.searchIconButton)} />}
            paperElevation={props.paperElevation}
          />
        </Grid>
      </Grid>
    </div>
  );
};

export default TableToolbar;