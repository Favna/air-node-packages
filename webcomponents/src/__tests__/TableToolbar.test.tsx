import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import TableToolbar from '../TableToolbar/TableToolbar';

let tableToolbar: ShallowWrapper;
const mockOnSearchInput = jest.fn();
const mockOnSearchClear = jest.fn();

beforeAll(() => tableToolbar = shallow(
  <TableToolbar
    searchplaceholderlabel='placeholder'
    onsearchinput={mockOnSearchInput} onsearchclear={mockOnSearchClear}
  />
));

test('should render without crashing', () => {
  expect(tableToolbar).toBeTruthy();
});

test('should trigger change when searching', () => {
  const inputBox = tableToolbar
    .find('[data-qa="table-search-bar"]')
    .dive().dive()
    .find('[data-qa="table-search-bar"]');

  const expectedCalledWith = { target: { value: 'DarthVader' } };
  const expectedResult = mockOnSearchInput(expectedCalledWith);

  expect(expectedResult).toBeUndefined();

  inputBox.simulate('change', expectedCalledWith);
  expect(mockOnSearchInput).toHaveBeenCalled();
  expect(mockOnSearchInput).toHaveBeenCalledWith(expectedCalledWith);
});

test('should trigger cancel when clicking X', () => {
  const cancelButton = tableToolbar
    .find('[data-qa="table-search-bar"]')
    .dive().dive()
    .find('WithStyles(ForwardRef(IconButton))')
    .last();

  cancelButton.simulate('click');
  expect(mockOnSearchClear).toHaveBeenCalled();
});

