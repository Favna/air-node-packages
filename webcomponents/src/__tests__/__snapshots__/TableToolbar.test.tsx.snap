// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className=""
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    justify="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      xs={6}
    />
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      xs={6}
    >
      <SearchBar
        className=""
        data-qa="table-search-bar"
        onCancelSearch={[MockFunction]}
        onChange={[Function]}
        placeholder="placeholder..."
        searchIcon={
          <Memo
            className=""
          />
        }
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
